File: scripts/collect_ratings.py
#!/usr/bin/env python3
"""Collecte automatique de notes critiques pour une liste de vins.

Usage:
    python3 scripts/collect_ratings.py --input wines.csv --output ratings.csv
"""

from __future__ import annotations

import argparse
import csv
import html
import json
import re
import time
from dataclasses import dataclass
from typing import Dict, Iterable, List
from urllib.parse import quote_plus
from urllib.request import Request, urlopen

USER_AGENT = (
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36"
)

RATING_PATTERNS: Dict[str, re.Pattern[str]] = {
    "robert_parker": re.compile(
        r"(?:Robert\s+Parker|Parker)[^\d]{0,25}(\d{2,3}(?:[\.,]\d+)?(?:\s*/\s*\d{2,3})?)",
        re.IGNORECASE,
    ),
    "hachette": re.compile(
        r"(?:Guide\s+Hachette|Hachette)[^\d]{0,25}(\d{1,2}(?:[\.,]\d+)?(?:\s*/\s*\d{1,2})?)",
        re.IGNORECASE,
    ),
    "wine_spectator": re.compile(
        r"(?:Wine\s+Spectator)[^\d]{0,25}(\d{2,3}(?:[\.,]\d+)?(?:\s*/\s*\d{2,3})?)",
        re.IGNORECASE,
    ),
    "jancis_robinson": re.compile(
        r"(?:Jancis\s+Robinson)[^\d]{0,25}(\d{1,2}(?:[\.,]\d+)?(?:\s*/\s*\d{2})?)",
        re.IGNORECASE,
    ),
}


@dataclass
class WineQuery:
    domaine: str
    appellation: str
    millesime: str

    def to_search_query(self) -> str:
        return f"site:millesima.fr {self.domaine} {self.appellation} {self.millesime}".strip()


def fetch_html(url: str, timeout: int) -> str:
    req = Request(url, headers={"User-Agent": USER_AGENT})
    with urlopen(req, timeout=timeout) as resp:
        return resp.read().decode("utf-8", errors="ignore")


def extract_links_from_ddg(html_doc: str, max_results: int) -> List[str]:
    # DuckDuckGo HTML SERP has links in <a ... class="result__a" ... href="...">
    pattern = re.compile(r'<a[^>]*class="[^"]*result__a[^"]*"[^>]*href="([^"]+)"', re.IGNORECASE)
    links: List[str] = []
    for href in pattern.findall(html_doc):
        unescaped = html.unescape(href)
        if "millesima.fr" in unescaped:
            links.append(unescaped)
        if len(links) >= max_results:
            break
    return links


def duckduckgo_links(query: str, timeout: int, max_results: int) -> List[str]:
    url = f"https://duckduckgo.com/html/?q={quote_plus(query)}"
    try:
        html_doc = fetch_html(url, timeout=timeout)
    except Exception:
        return []
    return extract_links_from_ddg(html_doc=html_doc, max_results=max_results)


def html_to_text(html_doc: str) -> str:
    cleaned = re.sub(r"<script[\\s\\S]*?</script>", " ", html_doc, flags=re.IGNORECASE)
    cleaned = re.sub(r"<style[\\s\\S]*?</style>", " ", cleaned, flags=re.IGNORECASE)
    cleaned = re.sub(r"<[^>]+>", " ", cleaned)
    cleaned = html.unescape(cleaned)
    return re.sub(r"\s+", " ", cleaned).strip()


def extract_ratings_from_text(text: str) -> Dict[str, str]:
    ratings: Dict[str, str] = {}
    for source, pattern in RATING_PATTERNS.items():
        match = pattern.search(text)
        if match:
            ratings[source] = match.group(1).replace(" ", "")
    return ratings


def scrape_wine_ratings(wine: WineQuery, timeout: int, max_results: int, sleep: float) -> Dict[str, str]:
    query = wine.to_search_query()
    candidates = duckduckgo_links(query=query, timeout=timeout, max_results=max_results)
    if not candidates:
        return {"best_url": "", "ratings_json": "{}", "status": "no-candidate-page"}

    for url in candidates:
        try:
            html_doc = fetch_html(url, timeout=timeout)
        except Exception:
            continue

        text = html_to_text(html_doc)
        ratings = extract_ratings_from_text(text)
        if ratings:
            return {
                "best_url": url,
                "ratings_json": json.dumps(ratings, ensure_ascii=False, separators=(",", ":")),
                "status": "ok",
            }
        time.sleep(sleep)

    return {"best_url": candidates[0], "ratings_json": "{}", "status": "no-rating-found"}


def read_wines(path: str) -> Iterable[WineQuery]:
    with open(path, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        required = {"domaine", "appellation", "millesime"}
        missing = required - set(reader.fieldnames or [])
        if missing:
            raise ValueError(f"Colonnes manquantes dans le CSV: {', '.join(sorted(missing))}")

        for row in reader:
            yield WineQuery(
                domaine=(row.get("domaine") or "").strip(),
                appellation=(row.get("appellation") or "").strip(),
                millesime=(row.get("millesime") or "").strip(),
            )


def write_output(path: str, rows: List[Dict[str, str]]) -> None:
    fieldnames = ["domaine", "appellation", "millesime", "best_url", "ratings_json", "status"]
    with open(path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Collecte des notes critiques de vins")
    parser.add_argument("--input", required=True, help="CSV input: domaine,appellation,millesime")
    parser.add_argument("--output", required=True, help="CSV output enrichi")
    parser.add_argument("--max-results", type=int, default=5, help="Nombre max de pages candidates")
    parser.add_argument("--timeout", type=int, default=15, help="Timeout HTTP en secondes")
    parser.add_argument("--sleep", type=float, default=1.2, help="Pause entre requêtes")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    rows: List[Dict[str, str]] = []

    for wine in read_wines(args.input):
        result = scrape_wine_ratings(
            wine=wine,
            timeout=args.timeout,
            max_results=args.max_results,
            sleep=args.sleep,
        )
        rows.append(
            {
                "domaine": wine.domaine,
                "appellation": wine.appellation,
                "millesime": wine.millesime,
                **result,
            }
        )

    write_output(args.output, rows)


if __name__ == "__main__":
    main()

File: src/main/java/com/ratemywine/CollectRatings.java
package com.ratemywine;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class CollectRatings {
    private static final String USER_AGENT =
            "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 " +
                    "(KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36";

    private static final Map<String, Pattern> RATING_PATTERNS = Map.of(
            "robert_parker", Pattern.compile("(?:Robert\\s+Parker|Parker)[^\\d]{0,25}(\\d{2,3}(?:[\\.,]\\d+)?(?:\\s*/\\s*\\d{2,3})?)", Pattern.CASE_INSENSITIVE),
            "hachette", Pattern.compile("(?:Guide\\s+Hachette|Hachette)[^\\d]{0,25}(\\d{1,2}(?:[\\.,]\\d+)?(?:\\s*/\\s*\\d{1,2})?)", Pattern.CASE_INSENSITIVE),
            "wine_spectator", Pattern.compile("(?:Wine\\s+Spectator)[^\\d]{0,25}(\\d{2,3}(?:[\\.,]\\d+)?(?:\\s*/\\s*\\d{2,3})?)", Pattern.CASE_INSENSITIVE),
            "jancis_robinson", Pattern.compile("(?:Jancis\\s+Robinson)[^\\d]{0,25}(\\d{1,2}(?:[\\.,]\\d+)?(?:\\s*/\\s*\\d{2})?)", Pattern.CASE_INSENSITIVE)
    );

    private CollectRatings() {}

    public static void main(String[] args) throws Exception {
        Map<String, String> cli = ArgParser.toMap(args, 0);
        String input = required(cli, "--input");
        String output = required(cli, "--output");
        int maxResults = Integer.parseInt(cli.getOrDefault("--max-results", "5"));
        int timeout = Integer.parseInt(cli.getOrDefault("--timeout", "15"));
        double sleep = Double.parseDouble(cli.getOrDefault("--sleep", "1.2"));

        List<WineQuery> wines = readWines(Path.of(input));
        List<Map<String, String>> rows = new ArrayList<>();

        for (WineQuery wine : wines) {
            Map<String, String> result = scrapeWineRatings(wine, timeout, maxResults, sleep);
            LinkedHashMap<String, String> row = new LinkedHashMap<>();
            row.put("domaine", wine.domaine);
            row.put("appellation", wine.appellation);
            row.put("millesime", wine.millesime);
            row.putAll(result);
            rows.add(row);
        }

        writeOutput(Path.of(output), rows);
    }

    private static String required(Map<String, String> map, String key) {
        String value = map.get(key);
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("Paramètre requis: " + key);
        }
        return value;
    }

    private static String fetchHtml(String url, int timeout) throws IOException, InterruptedException {
        HttpClient client = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL)
                .connectTimeout(Duration.ofSeconds(timeout)).build();
        HttpRequest req = HttpRequest.newBuilder(URI.create(url))
                .header("User-Agent", USER_AGENT)
                .timeout(Duration.ofSeconds(timeout))
                .GET().build();
        HttpResponse<byte[]> response = client.send(req, HttpResponse.BodyHandlers.ofByteArray());
        return new String(response.body(), StandardCharsets.UTF_8);
    }

    private static List<String> duckduckgoLinks(String query, int timeout, int maxResults) {
        String encoded = URLEncoder.encode(query, StandardCharsets.UTF_8);
        String url = "https://duckduckgo.com/html/?q=" + encoded;
        try {
            String html = fetchHtml(url, timeout);
            return extractLinksFromDdg(html, maxResults);
        } catch (Exception e) {
            return List.of();
        }
    }

    private static List<String> extractLinksFromDdg(String html, int maxResults) {
        Pattern pattern = Pattern.compile("<a[^>]*class=\"[^\"]*result__a[^\"]*\"[^>]*href=\"([^\"]+)\"", Pattern.CASE_INSENSITIVE);
        List<String> links = new ArrayList<>();
        Matcher m = pattern.matcher(html);
        while (m.find() && links.size() < maxResults) {
            String href = m.group(1).replace("&amp;", "&");
            if (href.contains("millesima.fr")) {
                links.add(href);
            }
        }
        return links;
    }

    private static String htmlToText(String html) {
        String cleaned = html.replaceAll("(?is)<script[\\s\\S]*?</script>", " ");
        cleaned = cleaned.replaceAll("(?is)<style[\\s\\S]*?</style>", " ");
        cleaned = cleaned.replaceAll("<[^>]+>", " ");
        cleaned = cleaned.replace("&nbsp;", " ").replace("&amp;", "&");
        return cleaned.replaceAll("\\s+", " ").trim();
    }

    private static Map<String, String> extractRatingsFromText(String text) {
        Map<String, String> ratings = new LinkedHashMap<>();
        for (Map.Entry<String, Pattern> e : RATING_PATTERNS.entrySet()) {
            Matcher m = e.getValue().matcher(text);
            if (m.find()) {
                ratings.put(e.getKey(), m.group(1).replace(" ", ""));
            }
        }
        return ratings;
    }

    private static Map<String, String> scrapeWineRatings(WineQuery wine, int timeout, int maxResults, double sleepSeconds) throws InterruptedException {
        String query = "site:millesima.fr " + wine.domaine + " " + wine.appellation + " " + wine.millesime;
        List<String> candidates = duckduckgoLinks(query.trim(), timeout, maxResults);
        if (candidates.isEmpty()) {
            return Map.of("best_url", "", "ratings_json", "{}", "status", "no-candidate-page");
        }

        for (String url : candidates) {
            try {
                String html = fetchHtml(url, timeout);
                String text = htmlToText(html);
                Map<String, String> ratings = extractRatingsFromText(text);
                if (!ratings.isEmpty()) {
                    return Map.of(
                            "best_url", url,
                            "ratings_json", toCompactJson(ratings),
                            "status", "ok"
                    );
                }
                Thread.sleep((long) (sleepSeconds * 1000));
            } catch (Exception ignored) {
                // ignore candidate failure and continue
            }
        }

        return Map.of("best_url", candidates.get(0), "ratings_json", "{}", "status", "no-rating-found");
    }

    private static String toCompactJson(Map<String, String> map) {
        StringBuilder sb = new StringBuilder("{");
        int i = 0;
        for (Map.Entry<String, String> e : map.entrySet()) {
            if (i++ > 0) {
                sb.append(',');
            }
            sb.append('"').append(jsonEscape(e.getKey())).append('"').append(':')
                    .append('"').append(jsonEscape(e.getValue())).append('"');
        }
        sb.append('}');
        return sb.toString();
    }

    private static String jsonEscape(String raw) {
        return raw.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private static List<WineQuery> readWines(Path path) throws IOException {
        List<WineQuery> wines = new ArrayList<>();
        try (BufferedReader reader = Files.newBufferedReader(path, StandardCharsets.UTF_8)) {
            String header = reader.readLine();
            if (header == null) {
                return wines;
            }
            String[] columns = header.split(",", -1);
            Map<String, Integer> idx = new HashMap<>();
            for (int i = 0; i < columns.length; i++) {
                idx.put(columns[i].trim(), i);
            }
            if (!idx.containsKey("domaine") || !idx.containsKey("appellation") || !idx.containsKey("millesime")) {
                throw new IllegalArgumentException("Colonnes manquantes dans le CSV: domaine, appellation, millesime");
            }

            String line;
            while ((line = reader.readLine()) != null) {
                String[] values = line.split(",", -1);
                wines.add(new WineQuery(
                        safeGet(values, idx.get("domaine")),
                        safeGet(values, idx.get("appellation")),
                        safeGet(values, idx.get("millesime"))
                ));
            }
        }
        return wines;
    }

    private static String safeGet(String[] values, int index) {
        if (index < 0 || index >= values.length) {
            return "";
        }
        return values[index].trim();
    }

    private static void writeOutput(Path output, List<Map<String, String>> rows) throws IOException {
        try (BufferedWriter writer = Files.newBufferedWriter(output, StandardCharsets.UTF_8)) {
            writer.write("domaine,appellation,millesime,best_url,ratings_json,status\n");
            for (Map<String, String> row : rows) {
                writer.write(csv(row.getOrDefault("domaine", "")) + ","
                        + csv(row.getOrDefault("appellation", "")) + ","
                        + csv(row.getOrDefault("millesime", "")) + ","
                        + csv(row.getOrDefault("best_url", "")) + ","
                        + csv(row.getOrDefault("ratings_json", "{}")) + ","
                        + csv(row.getOrDefault("status", "")) + "\n");
            }
        }
    }

    private static String csv(String value) {
        if (value.contains(",") || value.contains("\"") || value.contains("\n")) {
            return "\"" + value.replace("\"", "\"\"") + "\"";
        }
        return value;
    }

    private record WineQuery(String domaine, String appellation, String millesime) {}

    private static final class ArgParser {
        private ArgParser() {}

        private static Map<String, String> toMap(String[] args, int from) {
            Map<String, String> map = new HashMap<>();
            for (int i = from; i < args.length; i++) {
                String key = args[i];
                if (!key.startsWith("--") || i + 1 >= args.length) {
                    continue;
                }
                map.put(key, args[i + 1]);
                i++;
            }
            return map;
        }
    }
}
